File: maximal_rectangle.py
class Solution:
    # optimal bottom up dp approach
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        maxArea = 0
        maxHeights = [0] * len(matrix[0])
        minLeftBounds = [float('-inf')] * len(matrix[0])
        maxRightBounds = [float('inf')] * len(matrix[0])

        for row in range(len(matrix)):
            # compute max heights from each cell using previous row's max heights
            for col in range(len(matrix[row])):
                if matrix[row][col] == '1':
                    maxHeights[col] += 1
                else:
                    maxHeights[col] = 0

            minLeftBound, maxRightBound = 0, len(matrix[row]) - 1
            # compute minimum left bound of rectangle that maximizes height from each cell
            for col in range(len(matrix[row])):
                if matrix[row][col] == '1':
                    # current cell's rectangle left bound is either previous one above with
                    # tighter bound that maximizes height, or current left bound of first
                    # '1' cell seen if no tigher bound from above exists
                    minLeftBounds[col] = max(minLeftBound, minLeftBounds[col])
                else:
                    minLeftBounds[col] = float('-inf')
                    minLeftBound = col + 1

            # compute maximum right bound of rectangle that maximizes height from each cell
            for col in range(len(matrix[row]) - 1, -1, -1):
                if matrix[row][col] == '1':
                    # current cell's rectangle right bound is either previous one above with
                    # tighter bound that maximizes height, or current right bound of first
                    # '1' cell seen if no tigher bound from above exists
                    maxRightBounds[col] = min(maxRightBound, maxRightBounds[col])
                else:
                    maxRightBounds[col] = float('inf')
                    maxRightBound = col - 1

            # compute max rectangle area given max height and min left and max right rectangle
            # bounds from each cell
            for col in range(len(matrix[row])):
                if matrix[row][col] == '1':
                    curArea = maxHeights[col] * (maxRightBounds[col] - minLeftBounds[col] + 1)
                    maxArea = max(curArea, maxArea)

        return maxArea

File: cpp/maximal_rectangle.cpp
class Solution {
public:
    // optimal monotonic stack approach
    int maximalRectangle(vector<vector<char>> &matrix) {
        if (matrix.empty()) {
            return 0;
        }

        int maxArea = 0;
        vector<int> heightsAtRow(matrix[0].size(), 0);

        // compute height at each cell row by row, then compute max
        // histogram rectangle from each row
        for (int row = 0; row < matrix.size(); ++row) {
            for (int col = 0; col < matrix[row].size(); ++col) {
                if (matrix[row][col] == '0') {
                    heightsAtRow[col] = 0;
                }
                else {
                    ++heightsAtRow[col];
                }
            }
            maxArea = max(maxArea, largestRectangleArea(heightsAtRow));
        }

        return maxArea;
    }

    int largestRectangleArea(vector<int> &heights) {
        // stack stores monotonically increasing heights, defer computing area of
        // rectangle from it once a shorter height no longer accomondates
        stack<pair<int, int>> seenHeights;
        int maxArea = 0;

        for (int i = 0; i < heights.size(); ++i) {
            // assume rectangle at current bar height starts from current position
            int startPos = i;

            // reached smaller bar height, compute areas for previous taller rectangles
            // since we know they end here
            while (!seenHeights.empty() && seenHeights.top().first > heights[i]) {
                auto &[prevHeight, prevPos] = seenHeights.top(); seenHeights.pop();
                maxArea = max(prevHeight * (i - prevPos), maxArea);
                // rectangle at shorter current height could start from previous
                // taller height's position so carry it over
                startPos = prevPos;
            }

            seenHeights.push(make_pair(heights[i], startPos));
        }

        // pop out and compute rectangle areas from any remaining increasing heights
        while (!seenHeights.empty()) {
            auto &[prevHeight, prevPos] = seenHeights.top(); seenHeights.pop();
            int areaFromHeight = prevHeight * (heights.size() - prevPos);
            maxArea = max(areaFromHeight, maxArea);
        }

        return maxArea;
    }
};
